test/Makefile

```diff
@@ -61,151 +67,149 @@ $(WASM_TOOLS): | $(DOWNDIR)
+
 clean::
 	rm -rf $(DOWNDIR)
 
-##### BUILD ####################################################################
+##### INFRA ####################################################################
 
-# For now, we list out the tests that we can currently build and run. This is
-# heavily focused on the functional tests; in the future it would be good to
-# fill out the missing tests and also include some `src/api` and `src/math`
-# tests (TODO).
-TESTS := \
-	$(LIBC_TEST)/src/functional/argv.c \
-	$(LIBC_TEST)/src/functional/basename.c \
-	$(LIBC_TEST)/src/functional/clocale_mbfuncs.c \
-	$(LIBC_TEST)/src/functional/clock_gettime.c \
-	$(LIBC_TEST)/src/functional/crypt.c \
-	$(LIBC_TEST)/src/functional/dirname.c \
-	$(LIBC_TEST)/src/functional/env.c \
-	$(LIBC_TEST)/src/functional/fnmatch.c \
-	$(LIBC_TEST)/src/functional/iconv_open.c \
-	$(LIBC_TEST)/src/functional/mbc.c \
-	$(LIBC_TEST)/src/functional/memstream.c \
-	$(LIBC_TEST)/src/functional/qsort.c \
-	$(LIBC_TEST)/src/functional/random.c \
-	$(LIBC_TEST)/src/functional/search_hsearch.c \
-	$(LIBC_TEST)/src/functional/search_insque.c \
-	$(LIBC_TEST)/src/functional/search_lsearch.c \
-	$(LIBC_TEST)/src/functional/search_tsearch.c \
-	$(LIBC_TEST)/src/functional/snprintf.c \
-	$(LIBC_TEST)/src/functional/sscanf.c \
-	$(LIBC_TEST)/src/functional/strftime.c \
-	$(LIBC_TEST)/src/functional/string.c \
-	$(LIBC_TEST)/src/functional/string_memcpy.c \
-	$(LIBC_TEST)/src/functional/string_memmem.c \
-	$(LIBC_TEST)/src/functional/string_memset.c \
-	$(LIBC_TEST)/src/functional/string_strchr.c \
-	$(LIBC_TEST)/src/functional/string_strcspn.c \
-	$(LIBC_TEST)/src/functional/string_strstr.c \
-	$(LIBC_TEST)/src/functional/strtod.c \
-	$(LIBC_TEST)/src/functional/strtod_long.c \
-	$(LIBC_TEST)/src/functional/strtod_simple.c \
-	$(LIBC_TEST)/src/functional/strtof.c \
-	$(LIBC_TEST)/src/functional/strtol.c \
-	$(LIBC_TEST)/src/functional/swprintf.c \
-	$(LIBC_TEST)/src/functional/tgmath.c \
-	$(LIBC_TEST)/src/functional/udiv.c \
-	$(LIBC_TEST)/src/functional/wcsstr.c \
-	$(LIBC_TEST)/src/functional/wcstol.c
-
-# Part of the problem including more tests is that the `libc-test`
-# infrastructure code is not all Wasm-compilable. As we include more tests
-# above, this list will also likely need to grow.
-COMMON_TEST_INFRA = \
-	$(LIBC_TEST)/src/common/path.c \
+INFRA_OBJDIR := $(OBJDIR)/common
+$(INFRA_OBJDIR):
+	mkdir -p $@
+
+# Build the common test infrastructure. Part of the problem including more tests
+# is that the `libc-test` infrastructure code is not all Wasm-compilable. As we
+# include more tests above, this list will also likely need to grow.
+INFRA_FILES = $(LIBC_TEST)/src/common/path.c \
 	$(LIBC_TEST)/src/common/print.c \
 	$(LIBC_TEST)/src/common/rand.c \
 	$(LIBC_TEST)/src/common/utf8.c
+$(INFRA_FILES): $(DOWNLOADED)
+INFRA_WASM_OBJS := $(patsubst $(LIBC_TEST)/src/common/%.c,$(OBJDIR)/common/%.wasm.o,$(INFRA_FILES))
+$(OBJDIR)/common/%.wasm.o: $(LIBC_TEST)/src/common/%.c | $(INFRA_OBJDIR)
```

Comment:
What does the vertical bar here in the rule do?

Reply:
It's an [order-only prerequisite](https://www.gnu.org/software/make/manual/make.html#Prerequisite-Types): `make` ensures the prerequisite is built but it doesn't worry about when it has been updated; for some reason I've found this helpful with directories when I don't want to trigger a rebuild of the actual objects.